[Frontend] Add Virtual File System Support to the Frontend

CodaFi opened the pull request:
Adds the -vfsoverlay frontend option that allows the user to pass
VFS overlay YAML files to Swift. These files define a (potentially
many-layered) virtual mapping on which we predicate a VFS.

Switch all input-based memory buffer reads in the Frontend to the new
FileSystem-based approach.

CodaFi committed:
Plumbing for a Virtual File System

Adds the -vfsoverlay frontend option that enables the user to pass
VFS overlay YAML files to Swift. These files define a (potentially
many-layered) virtual mapping on which we predicate a VFS.

Switch all input-based memory buffer reads in the Frontend to the new
FileSystem-based approach.

CodaFi committed:
Introduce SourceManager::getDisplayNameForLoc

Add an abstraction to retrieve a name suitable for display for a given
source location for use with diagnostics, debug locations, and magic
file literals.

CodaFi commented on test/Frontend/vfs.swift:
```diff
@@ -0,0 +1,19 @@
+// The Clang Importer should inherit Swift's VFS
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -I %t -DTEST_VFS_CLANG_IMPORTER -import-objc-header %t/VFSMappedModule.h -vfsoverlay %t/overlay.yaml -typecheck %s
+
+#if TEST_VFS_CLANG_IMPORTER
```
Just now thought of testing debug info and #file.  Will add a follow-up commit.

CodaFi committed:
Diagnose missing and invalid vfs overlay files

davidungar reviewed (changes requested):
The code looks fine, but someone approaching it for the first time would really benefit from some more comments explaining the purpose of it all, including the reason to support VFS.  Maybe a Doxygen comment for setUpFileSystem? You have good comments explaining the what, but I'm asking for more explanation of the why.

jrose-apple commented on include/swift/Basic/SourceManager.h:
```diff
@@ -17,13 +17,16 @@
 #include "llvm/ADT/StringMap.h"
 #include "llvm/Support/SourceMgr.h"
+#include "clang/Basic/FileManager.h"
+#include "swift/Basic/FileSystem.h"
```
(here and elsewhere) The order here should go "swift", "clang", "llvm".

(Also, hopefully some day we can sink the Clang FileManager down to LLVM.)

jrose-apple commented on include/swift/Basic/SourceManager.h:
```diff
@@ -43,13 +52,24 @@ class SourceManager {
   mutable std::pair<const char *, const VirtualFile*> CachedVFile = {nullptr, nullptr};
 
 public:
+  SourceManager(llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem> FS = clang::vfs::getRealFileSystem())
```
Nitpick: 80 columns, please.

jrose-apple commented on lib/Basic/SourceLoc.cpp:
```diff
@@ -39,6 +39,26 @@ SourceLoc SourceManager::getCodeCompletionLoc() const {
+
+  // Populate the cache with a (virtual) stat.
+  if (auto Status = FileSystem->status(Ident)) {
+    return (StatusCache[Ident] = Status.get()).getName();
```
:-/ No need to be clever with this. `get()` is basically free.

jrose-apple commented on lib/Basic/SourceLoc.cpp:
```diff
@@ -39,6 +39,26 @@ SourceLoc SourceManager::getCodeCompletionLoc() const {
+  // Next, try the stat cache
+  auto Ident = getIdentifierForBuffer(findBufferContainingLoc(Loc));
+  if (StatusCache.count(Ident)) {
+    return StatusCache[Ident].getName();
```
Nitpick: You can avoid a lookup here by getting the iterator instead.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -459,8 +459,12 @@ static bool ParseSearchPathArgs(SearchPathOptions &Opts,
+
   for (const Arg *A : Args.filtered(OPT_L)) {
-    Opts.LibrarySearchPaths.push_back(resolveSearchPath(A->getValue()));
+    Opts.ImportSearchPaths.push_back(resolveSearchPath(A->getValue()));
```
This change is incorrect; please revert. (-L paths go to the frontend for immediate mode, not for import resolution.)

jrose-apple commented on lib/Frontend/Frontend.cpp:
```diff
@@ -166,6 +167,38 @@ bool CompilerInstance::setup(const CompilerInvocation &Invok) {
+void CompilerInstance::setUpFileSystem() {
+  auto BaseFS = clang::vfs::getRealFileSystem();
+  auto Overlay = llvm::IntrusiveRefCntPtr<clang::vfs::OverlayFileSystem>(
+                    new clang::vfs::OverlayFileSystem(BaseFS));
```
Can overlay options refer to files in previous overlays? It's okay if the answer is no, but it probably deserves a comment.

CodaFi replied:
This is something we should ask @vsapsai.  Currently, `clang` always uses the base filesystem to look up overlay files.

vsapsai replied:
There is a desire to allow file lookup in previous overlays rdar://problem/39440687 And as for me, chaining file systems should work better when we add VFS that doesn't fall back to underlying file system rdar://problem/39465552 I have a change for Clang that moves VFS in that direction but it's not entirely ready yet: not enough tests, some blatantly copy-pasted iterator code. I'll post a draft so that we can align our changes.

vsapsai replied:
And here is the promised Clang change https://reviews.llvm.org/D50032

jrose-apple replied:
(Even with that as the eventual direction I think it's okay to land this as is; just that we may have to go back and change it to layer on VFSs one by one.)

davidungar reviewed (approved):
LGTM!

davidungar commented:
There is some place where batch-mode gets a filename from a buffer. This change could possibly interact with that.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -951,6 +951,7 @@ ClangImporter::create(ASTContext &ctx,
   }
   auto &instance = *importer->Impl.Instance;
   instance.setInvocation(importer->Impl.Invocation);
+  instance.setVirtualFileSystem(ctx.SourceMgr.getFileSystem());
```
Ah, hang on. This will trash any `-ivfsoverlay` options that got passed through `-Xcc`, right? For backwards compatibility, can we have this only get set if there are Swift-level `-vfsoverlay` options present? (We have access to the SearchPathOptions through the Swift ASTContext, or you could check whether the SourceManager's FileSystem has any overlays set.)

davidungar replied:
Sounds like a comment is needed along with the fix, so folks who come by this later will have the benefit of @jrose-apple 's highi-level view.

CodaFi replied:
The plan is to declare `-vfsoverlay` incompatible with any current and future `-Xcc` VFS overlays options to avoid this ordering problem altogether and bake in the "we're nuking your overlays" behavior.  We'll warn about this with a diagnostic in the mean time - one that recommends migrating to -vfsoverlay

CodaFi committed:
Diagnose clobbering clang's VFS with our own

There are three modes of concern here:

1) User provides -vfsoverlay but not -Xcc -ivfsoverlay

Recommended behavior.

2) User provides -Xcc -ivfsoverlay but not -vfsoverlay

Legacy behavior.  Clang will handle setup for the VFS.

3) User provides both -vfsoverlay and -Xcc -ivfsoverlay

Unsupported.  We will ignore the clang VFS overlays and substitute our
own.  We will also pop a warning about this.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
Big changes need big guns

@swift-ci please test and merge

CodaFi commented:
@swift-ci please test and merge
